### ISSUE TITLE
Could the pandas progress bar be used for series objects?

### ISSUE THREAD

**Issue by owlas** [tqdm/tqdm#182]
If I have a very long series with an expensive function to apply, it would be good to use the pandas `.progess_apply` too. It seems that only dataframes and groupbys have this ability.

For example
```python
df = pd.DataFrame(range(1000), columns=['numbers'])
tqdm_pandas(tqdm_notebook())
df.numbers.progress_apply(lambda x: x**2)
```
Gives the following error:
``` 
AttributeError: 'Series' object has no attribute 'progress_apply'
```

**Comment by owlas** [comment]
Thanks, that looks like it will solve it.

I had one more question. I was using the following temporary solution in a jupyter notebook:
```python
def tqdm_pandas(t):
  from pandas.core.frame import Series
  def inner(series, func, *args, **kwargs):
      t.total = series.size
      def wrapper(*args, **kwargs):
          t.update(1)
          return func(*args, **kwargs)
      result = series.apply(wrapper, *args, **kwargs)
      t.close()
      return result
  Series.progress_apply = inner
```

but I got the following output using the jupyter progress bar:
```python
tqdm_pandas(tqdm_notebook())
df.progress_apply(something)
```
![screen shot 2016-06-16 at 17 23 26](https://cloud.githubusercontent.com/assets/11660098/16124487/13ae3330-33e7-11e6-9b2a-7362826a9352.png)

Do you have the same problem?

**Comment by casperdcl** [comment]
your code seems fine. I used len(series) instead of series.size - let me know which is better (I don't actually use pandas myself). Your issue now seems to be a notebook one... @lrq3000? (I don't actually use notebook either, ha)

**Comment by lrq3000** [comment]
I will take a look later today.

2016-06-17 9:09 GMT+02:00 Casper da Costa-Luis <notifications@github.com>:

**Comment by lrq3000** [comment]
Please excuse me for the big delay.

I took a look at the issue, it happens because `tqdm_pandas` takes an already instanciated bar object. Since `tqdm_notebook` is instanciated without a total, the bar gets initialized with a light blue color, a max limit of 1 and the weird `n/|/etc.` as a status message because `n/|/` is a custom bar_format trick used to provide an easy access to the current n value inside `status_printer.print_status` (which cannot directly access n because of scope since this is a nested function).

To summarize: once `tqdm_notebook` is initiated, there's no way to change the total. So either:

* `tqdm_notebook.status_printer` needs to be changed so that it regularly checks the total and updates the bar style (which would lead to decreased performance and problems with error management or keyboard interrupt from IPython).

* or `tqdm_pandas` needs to use a bar class instead of a bar object instance, so that it can instanciate `tqdm_notebook` by itself after precomputing the dataframe total.

I chose to implement the second option, because I think it is the most reliable choice for the future (I guess tqdm_gui may suffer from a similar issue, and future more complex submodules will obviously suffer from the same problem also, because complex modules usually have complex init functions).

However, this breaks retrocompatibility.

Please [try the fix here](https://github.com/tqdm/tqdm/tree/pandas_notebook_fix) and please report if this fixes all your issues @owlas.

**Comment by casperdcl** [comment]
hm I need to look at this. was pretty sure total was not cached when updating for 'normal' tqdm, don't see how there would be a perf hit on gui/notebook.

**Comment by lrq3000** [comment]
@casperdcl No no for core tqdm the total is not cacher in status_printer, it's only necessary for the tqdm_notebook module because we instanciate a widget bar object and stick with it until the end, whereas tqdm redraws the whole display when required to. Indeed, core tqdm does not need to instanciate anything display-wise, the only thing that is initiated is the internal variables states, but nothing related directly to the display. If we were using curses, it would maybe be different.

For tqdm_notebook, we need to instanciate a widget bar, so the init is important and cannot be reactivated (it could but we would have to make a special function to first destroy the widget, then recreate it, etc., plus there is a bug currently in ipywidgets in the management of widgets deletion). So that's why I guess that tqdm_gui may suffer from a similar issue, but it depends on how it works (I admit I don't understand the details!).

About the perf hit, no it would be only for tqdm_notebook because the alternative solution would either have to a) parse the string to get not only n but the total, and dynamically change the bar max, b) make print_status and status_printer normal methods to be able to access self.total.

**Comment by lrq3000** [comment]
/EDIT: another solution would be to instanciate the ipywidget bar before the status_printer, and then provide it with the object reference. Then we could update the bar max from outside of status_printer, just like tqdm does.

I'm not sure this would work, but this is an idea.

**Comment by casperdcl** [comment]
original issue fixed in #183, new issue discussion in #192
